external/fv3fit/tests/training/test_main.py

```diff
@@ -0,0 +1,224 @@
+import dataclasses
+from typing import Optional, Sequence
+import fv3fit
+import fv3fit.train
+from fv3fit._shared.io import register
+import yaml
+import pytest
+import os
+import numpy as np
+import cftime
+import xarray as xr
+import loaders
+import vcm
+from unittest import mock
+
+
+@pytest.fixture
+def mock_dataset():
+
+    n_x, n_y, n_z, n_tile, n_time = (8, 8, 10, 6, 9)
+    arr = np.zeros((n_time, n_tile, n_z, n_y, n_x))
+    arr_surface = np.zeros((n_time, n_tile, n_y, n_x))
+    dims = ["time", "tile", "z", "y", "x"]
+    dims_surface = ["time", "tile", "y", "x"]
+
+    data = xr.Dataset(
+        {
+            "specific_humidity": (dims, arr),
+            "air_temperature": (dims, arr),
+            "downward_shortwave": (dims_surface, arr_surface),
+            "net_shortwave": (dims_surface, arr_surface),
+            "downward_longwave": (dims_surface, arr_surface),
+            "dQ1": (dims, arr),
+            "dQ2": (dims, arr),
+            "dQu": (dims, arr),
+            "dQv": (dims, arr),
+        },
+        coords={
+            "time": [
+                cftime.DatetimeJulian(2016, 8, day) for day in range(1, 1 + n_time)
+            ]
+        },
+    )
+
+    return data
+
+
+@dataclasses.dataclass
+class MainArgs:
+    data_path: str
+    train_config: str
+    train_data_config: str
+    val_data_config: str
+    output_path: str
+    local_download_path: Optional[str] = None
+
+
+class MockHyperparameters:
+    param1: str = ""
+
+
+@pytest.fixture
+def mock_train_dense_model():
+    train_mock = mock.MagicMock(name="train_dense_model")
+    train_mock.return_value = mock.MagicMock(
+        name="train_dense_model_return", spec=fv3fit.Predictor
+    )
+    register("mock")(train_mock.return_value.__class__)
+    original_func = fv3fit.get_training_function("DenseModel")
+    try:
+        fv3fit._shared.config.register_training_function(
+            "DenseModel", fv3fit.DenseHyperparameters
+        )(train_mock)
+        yield train_mock
+    finally:
+        fv3fit._shared.config.register_training_function(
+            "DenseModel", fv3fit.DenseHyperparameters
+        )(original_func)
+        register._model_types.pop("mock")
+
+
+@pytest.fixture
+def mock_load_batches():
+    magic_load_mock = mock.MagicMock(name="load_batches")
+    with mock.patch.object(loaders.BatchesConfig, "load_batches", magic_load_mock):
+        yield magic_load_mock
+
+
+@pytest.mark.parametrize("additional_variables", [[], ["downward_shortwave"]])
+def test_main(
```

Comment:
Yeah, I could do that. I originally didn't because "training takes long" but in this case the training codes are all mocks, so it should still be fast to call main 5x as many times. I'll try to merge in the "with_derived_output_variables" test by parameterizing over its value, and having a check that only occurs if it is non-empty.